califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
     return kSUCCESS;
 }
 
-bool R3BCalifaCrystalCal2Cluster::Match(R3BCalifaCrystalCalData* ref, R3BCalifaCrystalCalData* hit)
+void R3BCalifaCrystalCal2Cluster::Exec(Option_t* opt)
 {
-    if (ref == hit)
-        return 1;
-
-    auto circleAbs = [](double dphi) {
-        double d = fmod(fabs(dphi), 2 * M_PI);
-        return d < M_PI ? d : 2 * M_PI - d;
-    };
-    // Clusterization: you want to put a condition on the angle between the highest
-    // energy crystal and the others. This is done by using the TVector3 classes and
-    // not with different DeltaAngle on theta and phi, to get a proper solid angle
-    // and not a "square" one.                    Enrico Fiori
-    TVector3 vref = this->GetAnglesVector(ref->GetCrystalId());
-    TVector3 vhit = this->GetAnglesVector(hit->GetCrystalId());
-    bool takeCrystalInCluster = false;
-
-    // Check if the angle between the two vectors is less than the reference angle.
-    switch (fClusterAlgorithmSelector)
+    Reset();
+
+    const int numCrystalHits = fCrystalCalData->GetEntries();
+
+    R3BLOG(debug1, "Crystal hits at start:" << numCrystalHits);
+
+    vector<R3BCalifaCrystalCalData*> allCrystalVec;
+    vector<R3BCalifaCrystalCalData*> protonCandidatesVec;
+    vector<R3BCalifaCrystalCalData*> gammaCandidatesVec;
+    vector<R3BCalifaCrystalCalData*> saturatedCandidatesVec;
+
+    vector<Int_t> usedCrystals;
+
+    Double_t cryEnergy;
+    Int_t cryId;
+
+    for (Int_t i = 0; i < numCrystalHits; i++)
     {
-        case RECT:
-        { // rectangular window
-            if (TMath::Abs(vref.Theta() - vhit.Theta()) < fDeltaPolar &&
-                circleAbs(vref.Phi() - vhit.Phi()) < fDeltaAzimuthal)
+
+        cryId = dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i))->GetCrystalId();
+        cryEnergy = dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i))->GetEnergy();
+
+        if (cryEnergy >= fCrystalThreshold)
+            allCrystalVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
```

Comment:
I think both options are similar ?

Reply:
Basically yes. Currently the hits below the threshold will go through the lines till the end of the for loop although they would be rejected anyway. I added this comment just in case if someone adding a modification within the loop would forget adding a energy condition resulting in some unexpected behaviour. But yes, basically that would be fine as it is.